js/ethereum-utils/src/signature.ts

```diff
@@ -0,0 +1,497 @@
+
+  return {
+    type: 'ItemizedSignaturePayloadV2',
+    schemaId,
+    targetHash,
+    expiration,
+    actions,
+  };
+}
+
+/**
+ * Build a PaginatedDeleteSignaturePayloadV2 for signing.
+ *
+ * @param schemaId   uint16 schema identifier
+ * @param pageId     uint16 page identifier
+ * @param targetHash uint32 page hash
+ * @param expiration uint32 expiration block
+ */
+export function createPaginatedDeleteSignaturePayloadV2(
+  schemaId: number,
+  pageId: number,
+  targetHash: number,
+  expiration: number
+): PaginatedDeleteSignaturePayloadV2 {
+  assert(isValidUint16(schemaId), 'schemaId should be a valid uint16');
+  assert(isValidUint16(pageId), 'pageId should be a valid uint16');
+  assert(isValidUint32(targetHash), 'targetHash should be a valid uint32');
+  assert(isValidUint32(expiration), 'expiration should be a valid uint32');
+
+  return {
+    type: 'PaginatedDeleteSignaturePayloadV2',
+    schemaId,
+    pageId,
+    targetHash,
+    expiration,
+  };
+}
+
+/**
+ * Build a PaginatedUpsertSignaturePayloadV2 for signing.
+ *
+ * @param schemaId   uint16 schema identifier
+ * @param pageId     uint16 page identifier
+ * @param targetHash uint32 page hash
+ * @param expiration uint32 expiration block
+ * @param payload    HexString or Uint8Array data to upsert
+ */
+export function createPaginatedUpsertSignaturePayloadV2(
+  schemaId: number,
+  pageId: number,
+  targetHash: number,
+  expiration: number,
+  payload: HexString | Uint8Array
+): PaginatedUpsertSignaturePayloadV2 {
+  const parsedPayload: HexString = typeof payload === 'object' ? u8aToHex(payload) : payload;
+
+  assert(isValidUint16(schemaId), 'schemaId should be a valid uint16');
+  assert(isValidUint16(pageId), 'pageId should be a valid uint16');
+  assert(isValidUint32(targetHash), 'targetHash should be a valid uint32');
+  assert(isValidUint32(expiration), 'expiration should be a valid uint32');
+  assert(isValidHexString(parsedPayload), 'payload should be valid hex');
+
+  return {
+    type: 'PaginatedUpsertSignaturePayloadV2',
+    schemaId,
+    pageId,
+    targetHash,
+    expiration,
+    payload: parsedPayload,
+  };
+}
+
+/**
+ * Returns the EIP-712 browser request for a AddKeyData for signing.
+ *
+ * @param msaId           MSA ID (uint64) to add the key
+ * @param newPublicKey    32 bytes public key to add in hex or Uint8Array
+ * @param expirationBlock Block number after which this payload is invalid
+ * @param domain
+ */
+export function getEip712BrowserRequestAddKeyData(
+  msaId: string | bigint,
+  newPublicKey: HexString | Uint8Array,
+  expirationBlock: number,
+  domain: EipDomainPayload = EIP712_DOMAIN_DEFAULT
+): unknown {
+  const message = createAddKeyData(msaId, newPublicKey, expirationBlock);
+  const normalized = normalizePayload(message);
+  return {
+    types: {
+      ...EIP712_DOMAIN_DEFINITION,
+      ...ADD_KEY_DATA_DEFINITION,
+    },
+    primaryType: message.type,
+    domain: domain,
+    message: normalized,
+  };
+}
```

Comment:
nit: this seems like a good refactoring target, since it's duplicated for every payload type. Perhaps:
```js
function createEip712Payload(domain: EipDomainPayload, primaryType: any, message: any, data: any) {
   return {
      types: {
         ...EIP712_DOMAIN_DEFINITION,
         ...data,
      },
      primaryType,
      domain,
      message,
   };
}
```

Reply:
applied